### ISSUE TITLE
Why the lstm is not built using the container?

### ISSUE THREAD

**Issue by nosyndicate** [karpathy/char-rnn#136]
Hi, this may be a silly question since I am new with torch and lstm. I noticed that in LSTM.lua, the code builds the entire network, instead of building a lstm layer. Is this different from building a single layer of LSTM and use nn.Sequential to stack them all together?

**Comment by kmnns** [comment]
You mean this one, right? <https://github.com/Element-Research/rnn>
I think this module was not out of beta until later in 2015.

I am thinking about rewriting the mess I have made out of this codebase to use the official RNN module, in the hope of getting rid of this problem:

https://github.com/karpathy/char-rnn/issues/108

Large sequences blow up the GPU memory because something weird is going on with the cloning. This is a show stopper for me.

**Comment by nosyndicate** [comment]
@kmnns 
Thanks for the reply. So I guess the lstm.lua is a quick implementation right? At first I was worried that there is a specific reason of not doing in layer-wise way.

**Comment by kmnns** [comment]
I cannot speak for the author, but I guess the reason is just historic.

I am currently working on moving over to the official LSTM.lua container. The reason is epic problems with memory consumption on long sequences. It seems that this code's RNN memory consumption increases linearly with the sequence length.

Which I don't quite understand. The parameters between the clones seem to be shared (see the linked issue).

Don't know if the official container is any better. It also uses clones to hack recurrence.

**Comment by hughperkins** [comment]
What do you mean by 'official LSTM.lua' container?  Do you mean element-research?  (Just checking, since I want to make sure I'm considering all possible potential options for my own tasks)

**Comment by hughperkins** [comment]
> Don't know if the official container is any better. It also uses clones to hack recurrence.

Presumably we have to store the outputs and gradients somewhere.  If they fit in GPU memory that's good, otherwise we'll have to shuffle them backwards and forwards between main memory and GPU, or in worst case, save to disk...  But I'm not sure if sequences longer than 100 are useful anyway, since the gradients are going to decay fairly quickly, or, at least, the training 'information' provided by one single backprop is going to get shared out over so many weights, that it will get kind of 'diluted' somehow?  I havent really experimented with this yet (at all really...)  Just kind of thinking out loud mostly :-P

**Comment by kmnns** [comment]
Yes element-research, because last time I checked it was officially endorsed by the Torch guys. (forgot where though, at some RNN example it says "deprecated, go to element-research")
Plus they are considering integrating it into the package: https://github.com/torch/torch7/blob/master/ROADMAP.md

**Comment by hughperkins** [comment]
> Yes element-research, because last time I checked it was officially endorsed by the Torch guys. (forgot where though, at some RNN example it says "deprecated, go to element-research")

Ok

> I am using this right now and don't have the memory problem anymore for long sequences and large layers.

Cool.  Good info.  Thanks! :-)

**Comment by wrapperband** [comment]
Sounds like this is my stopper also, is the code updatable? (small enough for contributer). Or is this code just a "seed package", and you've got to be a programmer still to do ML research?

**Comment by hughperkins** [comment]
@wrapperband You'd better have some scripting knowledge I reckon :-)

@karpathy I reckon this issue is a question that was answered in kmmns's reply at https://github.com/karpathy/char-rnn/issues/136#issuecomment-163240062 , and can be closed?
